src/coreclr/jit/importercalls.cpp

```diff
@@ -2601,6 +2601,33 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Runtime_InteropService_MemoryMarshal_GetArrayDataReference:
+            {
+                assert(sig->numArgs == 1);
+
+                GenTree*             array    = impPopStack().val;
+                CORINFO_CLASS_HANDLE elemHnd  = sig->sigInst.methInst[0];
+                CorInfoType          jitType  = info.compCompHnd->asCorInfoType(elemHnd);
+                var_types            elemType = JITtype2varType(jitType);
+
+                if (fgAddrCouldBeNull(array))
```

Comment:
Can you please explain why this part is needed? and why GTF_INX_ADDR_NONNULL is used

Reply:
`gtNewNullCheck` asserts if `fgAddrCouldBeNull` is false. `GTF_INX_ADDR_NONNULL` tells the JIT this won't be null since we check it separately ourselves.